sdk/core/Azure.Core/src/Shared/ClientDiagnostics.cs

```diff
@@ -62,14 +62,14 @@ internal static HttpMessageSanitizer CreateMessageSanitizer(DiagnosticsOptions d
 
         internal static string? GetResourceProviderNamespace(Assembly assembly)
         {
-            foreach (var customAttribute in assembly.GetCustomAttributes(true))
+            foreach (var customAttribute in assembly.GetCustomAttributesData())
             {
                 // Weak bind internal shared type
-                var attributeType = customAttribute.GetType();
-                if (attributeType.Name == "AzureResourceProviderNamespaceAttribute")
+                var attributeType = customAttribute.AttributeType!;
+                if (attributeType.FullName == ("Azure.Core.AzureResourceProviderNamespaceAttribute"))
                 {
-                    return attributeType.GetProperty("ResourceProviderNamespace")?.GetValue(customAttribute) as string;
-                }
+                    var namedArguments = customAttribute.ConstructorArguments;
+                    return namedArguments.Single().Value as string;                }
```

Comment:
Why can't you cast the attribute instance to the type?
Do you really require the ability for the type to be declared in any assembly?

Reply:
I'm not entirely sure what you mean by this question, since I don't have all of the context in this area, but I believe the answer to your second question is yes.
[documentation](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Shared/AzureResourceProviderNamespaceAttribute.cs):
```
/// This attribute should be set on all client assemblies with value of one of the resource providers
/// from the https://docs.microsoft.com/azure/azure-resource-manager/management/azure-services-resource-providers list.
``` 
examples of usage: https://grep.app/search?q=AzureResourceProviderNameSpace&filter[repo][0]=Azure/azure-sdk-for-net&filter[path][0]=sdk/
This attribute provides the resource provider name, this particular usage of the name is to use in our distributed tracing.

Reply:
Resolving this because no response.

Reply:
Sorry - I was on vacation last week.

The question is, why do you have to look for the attribute by its name. Isn't the attribute type defined in the core assembly as well? So why not just get an instance of the attribute, cast it to the `AzureResourceProviderNamespaceAttribute` and access the `ResourceProviderNamespace` property on it? It would be even easier with something like:
```C#
var attributes = assembly.GetCustomAttributes(typeof(AzureResourceProviderNamespaceAttribute), true);
if (attributes.Length > 0)
    return ((AzureResourceProviderNamespaceAttribute)attributes[0]).ResourceProviderNamespace;
```

Typically the only use case for looking up attributes by name is when the attribute is allowed to be defined in whatever assembly it wants to be. It might be the case here, I don't know.

Reply:
No worries! I'm unresolving this to see if anyone else has any opinions on this. My understanding after investigating is that it is technically defined in the core assembly but since it's part of our shared source we'd have to add a dependency on that type. It's easier to reference it indirectly.

Reply:
OK - makes sense. There are not trimming/AOT concerns around this, I was just curious about the added complexity. Thanks for looking into it.